### ISSUE TITLE
Maps: Accuracy of household GPS coordinates

### ISSUE THREAD

**Issue by gmcressman** [ICTatRTI/coconut-analytics#155]
So many households in the water is disturbing.

![image](https://cloud.githubusercontent.com/assets/329768/16561561/d4a626ae-41c7-11e6-8a93-fd6a8297e87e.png)

At least some of this is probably due to very poor registration of household location to the administrative boundary layers and the base map. @workmaps, I still think that we could improve alignment of the administrative layers to the base map by just taking one or more control points, calculating the offsets, and applying those offsets, dynamically, to coordinates of the admin layers. Household locations are much more difficult. We could take control points in the field and look at their offsets to derive offsets for the household point layer. In general, the poor alignment of the layers makes it difficult to use this as a tool for accurately identifying small groups of households that, for example, need IRS. I believe we can do better, but we need to consider this carefully. There are also issues in the accuracy of household locations.

I looked the HouseholdLocation-accuracy of the first 3902 records in the database. I already had an export of this, so I just worked with that. 

[Household_location_analysis.xlsx](https://github.com/ICTatRTI/coconut-analytics/files/346604/Household_location_analysis.xlsx)

@mikeymckay, please look at the few errors I've highlighted in the data. Determine whether any changes in the software could prevent some of these.

About 87% of the records contain GPS location data.  Based on the numbers, I assume GPS accuracy is expressed in meters. There is an enormous range in GPS accuracy, from 3 meters to 5 thousand. The mean is about 566 meters, which is not good. The standard deviation is over 1 thousand meters.

![image](https://cloud.githubusercontent.com/assets/329768/16566363/a2ad7aca-41e1-11e6-87c4-29536e2d6b36.png)

Accuracy does not appear to have changed over time.

![image](https://cloud.githubusercontent.com/assets/329768/16566372/c00d7714-41e1-11e6-8003-f5ad3321bab0.png)

Accuracy of household location is very important to the usefulness of this tool. Could this be improved through training? Could we change the software to warn the user when location accuracy was below some threshold? Other ideas?

**Comment by mikeymckay** [comment]
Are you using data from cococloud.co? I have added a random offset to location data on that server to anonymize it, so that may be one issue with households. Check the same map on the live system and compare. Accuracy has been a problem. We've done a couple of things:
1) Train people to make sure they have a clear path to the sky when capturing location
2) Added validation that warns them if the accuracy is poor (could make this an error which would force them to keep trying until they get it)
3) Added code that forces high accuracy mode
4) Started location seeking as soon as the form loads, not just when the button is pressed (to increase the likelihood of more satellites being found)

**Comment by gmcressman** [comment]
My analysis used an April 26 export of data from the production system. I used data from the following file:

Household-2014-12-29-2016-01-03-20160426-11481-pwqx9t.csv

-gmc

Gordon M. Cressman
Senior Program Director | Research Computing Division | RTI International
Email: gmc@rti.org<mailto:gmc@rti.org> | Office: +1 919 541-6363 | Mobile: +1 919 271-7003 | Skype: gmcressman

Was this email too brief? Here is why<http://emailcharter.org/>

**Comment by mikeymckay** [comment]
The map you made showing households in the water used the same data, or was
that from cococloud.co?

**Comment by gmcressman** [comment]
@mikeymckay, the map image from the new map interface, showing households in the water, is using the de-identified data. The offsets that you introduced could be causing at least part of this.

Here we have week 15 - week 21 of 2015, using the old map interface:

![image](https://cloud.githubusercontent.com/assets/329768/16567676/424abbc6-41ef-11e6-90af-fc34be30517f.png)

No points in the water. 

Over on the new map interface, for the same time period, I'm staring at a spinner, and am still waiting for the data query to complete. Why is the newer system so much slower? Something is not right with this.

**Comment by mikeymckay** [comment]
Thanks for looking into this @gmcressman. I don't think average is a good indicator of the mean in this instance, as outliers can really skew the mean. Median returns a mean of 20 meters for the dataset that you sent. We currently display an error when accuracy is over 50 meters, but I don't believe this error stops the user from successfully saving the record, it just encourages them to try again. 

I think the next step to improve accuracy with our current hardware would be to require multiple locations to be recorded, and then average those weighted by accuracy. It wouldn't be hard to do.

Here's the library we use to increase accuracy, it has some good discussion about constraints and how it all works: https://github.com/gwilson/getAccurateCurrentPosition

**Comment by gmcressman** [comment]
I agree, the median is 20 meters. I had meant to calculate that. The extreme tail to the right skews the average strongly to the positive. I wonder whether the coordinates are useful at all if accuracy is above a certain threshold, such as 50 meters. We could, of course, require the user to try again until accuracy is below a threshold. You’ve not done that, and may have good reasons not to.

The discussion at getAccurateCurrentPosition is useful.

The offset coordinates in the de-identified dataset make location accuracy appear much worse than it actually is. That is assuming this is responsible for the large numbers of households in the water. That could be as difficult to explain as explaining why we would make precise household location data available publicly. Since the PII/PHI data in each record have been de-identified, do the original household locations reveal any PII/PHI?

**Comment by mikeymckay** [comment]
Not sure what your experiences are with GPS, but it sometimes it takes a long time to get a good satellite fix. My assumption was that if it takes a long time to get a good satellite fix then it will take even longer to get good accuracy. This may not be a good assumption, and dropping the threshold to 30 meters or even 10 might improve things without taking much more time. And of course, we have already spent a lot of time, fuel, money, etc to get someone to the household - even if it takes 30 minutes to get a very accurate location it might be worth it. 50 meters is my fairly arbitrary choice determined after some informal testing.

In terms of the de-identified dataset, the locations have probably been re-de-identified at least 10 times, meaning that they are probably quite far from their actual location. This could be fixed by wiping the database, replicating a clean copy in, and then running the de-identified script again which should only move the locations by a few hundred meters if I remember correctly.

**Comment by gmcressman** [comment]
I’ve used handheld GPS units to get around the Western Desert of Egypt, and the Sinai. I trained our staff in Iraq to use handheld GPS units for navigation. The readme at getAccurateCurrentPosition rings true to my experience: 1) it can take 15 minutes or so to acquire 4 satellites in an a new area; 2) most GPS chips remember the positions of satellites, so reacquiring them in the same general area can take only seconds. The receiver needs a minimum of 4 satellites to compute position and the clock correction. It an potentially acquire at least 10 of the 31-satellite constellation, depending on the position of the receiver, the rotation of the earth, and the positions of the satellites. 

50 meters might be a practical threshold. 10 is too low, I believe, considering that our median accuracy now is about 20 meters. 

In the mapping display, we could just filter out accuracy below a threshold. We should also filter out records that do not contain LAT and LON, and records with non-numeric data in those fields. Is there any filtering in that query now?

It does look bad to talk about precise targeting of interventions, while showing so many households in the water. At some point the script should be run again, perhaps with smaller offsets.

While looking at actual locations in the older map interface, it struck me that the few points out in the water appeared to be aligned with ferry routes. That looks suspicious!

**Comment by workmaps** [comment]
Hi Mike. I am getting cases the couchdb instance using the Reports.getCases method.

**Comment by mikeymckay** [comment]
@workmaps It will be particularly slow during the first query when the data is not indexed. After that it should be fast.

@gmcressman no filtering is done. Of course, if there is no lat/lon it won't appear on the map. It should be easy for @workmaps to add a UI filter to remove data above a certain accuracy threshold. In terms of the ferry paths I don't think it is particularly meaningful - I doubt that those paths are accurate or that ferries even follow them as tides and currents change so much. I can refresh the data and then make sure we don't skew our locations too much.

**Comment by gmcressman** [comment]
@mikeymckay : On the flight to DC yesterday evening I discussed GPS accuracy with Bill Wheaton. He thought a mean accuracy of 20 meters was good. He quizzed me on how the readings were being taken, and how accuracy was being determined. He didn't have any suggestions for improving this. We discussed building a little more coaching into the app. Perhaps it has this already, but if the accuracy be below minimum threshold (say 50m), then we warn the user, and ask the user to try again after moving to the nearest location with a clear view of the sky. We could also explain to the user that getting an accurate location is important, and that it could take up to say, 15 minutes. You have a lot of recent experience in the field with this, so I'll leave this to your better judgement.

@workmaps : Bill thought it was a good idea to filter out points with accuracy below a set threshold, say 500m, when querying the data for the map displaly. This should be easy to do. He agreed that that this threshold should be user-configurable. @scloo , that would mean adding this to the user settings, and setting it to some default value.

Bill did have one more suggestion. He said that for the GATS survey, which has been run in many countries using RTI Mobile FS, they downloaded PDF satellite image tiles of the survey area onto the mobile device. The user was able to than view current GPS position overlaid on the image, and then drag the position marker over the image of the structure. This improved the correspondence between the household locations and the satellite imagery. The PDF images are small, but the images that correspond to the user's assigned data collection area must be downloaded onto the mobile device. All this is certainly out of scope of the current SIF. However, we've discussed this idea before, and it could have other uses. We should consider putting it in the backlog.

**Comment by workmaps** [comment]
Threash hold level to 50 meters as a setting variable.
